### PR TITLE
SchemaEncoder: encode schemas as descriptor.proto describes

### PR DESCRIPTION
**Pull request by swankjesse**
This is the first step towards fully supporting schema reflection.

Options are not yet implemented.

### CODE REVIEW

File: wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/internal/SchemaEncoder.kt [hunk 1]
```diff
@@ -0,0 +1,355 @@
+package com.squareup.wire.schema.internal
```
Comment (oldergod) [inline wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/internal/SchemaEncoder.kt:1]: copyrights and for the others as well still needed for wire I think

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/SchemaEncoderInteropTest.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2020 Square Inc.
```
Comment (pforhan) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/SchemaEncoderInteropTest.kt:2]: 2021

File: wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/internal/SchemaEncoder.kt [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    }
+  }
+
+  private val messageEncoder = object : Encoder<MessageType>() {
+    override fun encode(writer: ProtoWriter, value: MessageType) {
+      val syntax = schema.protoFile(value.type)!!.syntax
+
+      val syntheticMaps = collectSyntheticMapEntries(value.type.toString(), value.declaredFields)
+
+      // TODO(jwilson): can oneofs be maps?
```
Comment (oldergod) [inline wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/internal/SchemaEncoder.kt:85]: No.

**Review by oldergod — APPROVED** [review]
Super nice

**Pull request merged**